sources/sqlite3_cql_extension/cqlsqlite3extension.py

```diff
@@ -0,0 +1,355 @@
+    in_arguments = [arg for arg in proc['args'] if arg.get('binding', 'in') == ('in')]
+    inout_arguments = [arg for arg in proc['args'] if arg.get('binding', 'in') == ('inout')]
+    out_arguments = [arg for arg in proc['args'] if arg.get('binding', 'in') == ('out')]
+
+    innie_arguments = in_arguments + inout_arguments
+    outtie_arguments = inout_arguments + out_arguments
+
+
+    print(f"void {proc['canonicalName']}(sqlite3_context *_Nonnull context, int32_t argc, sqlite3_value *_Nonnull *_Nonnull argv)", end="")
+    print("{")
+
+    print(f"  // 1. Ensure Sqlite function argument count matches count of the procedure in and inout arguments")
+    print(f"  if (argc != {len(innie_arguments)}) goto invalid_arguments;")
+    print(f"")
+
+
+    print("  // 2. Ensure sqlite3 value type is compatible with cql type")
+    for index, arg in enumerate(innie_arguments):
+        print(
+            f"  if (!is_sqlite3_type_compatible_with_cql_core_type("
+                f"sqlite3_value_type(argv[{index}]), "
+                f"{row_types[arg['type']]}, "
+                f"{'cql_false' if arg['isNotNull'] else 'cql_true'}"
+            f")) goto invalid_arguments; // {arg['name']}"
+        )
+    print("")
+
+
+    print("  // 3. Marshalled argument initialization")
+    for index, arg in enumerate(innie_arguments):
+        if arg['type'] in ("text", "blob", "object"):
+            print(f"  {cql_nullable_types[arg['type']].ljust(20)} {arg['name'].ljust(25)} = {sqlite3_value_getter_nullable[arg['type']]}(argv[{index}]);")
+        elif arg['isNotNull']:
+            print(f"  {cql_notnull_types[arg['type']].ljust(20)} {arg['name'].ljust(25)} = {sqlite3_value_getter_notnull[arg['type']]}(argv[{index}]);")
+        else:
+            print(f"  {cql_nullable_types[arg['type']].ljust(20)} {arg['name'].ljust(25)} = {sqlite3_value_getter_nullable[arg['type']]}(argv[{index}]);")
+    for arg in out_arguments:
+        if arg['type'] in ("text", "blob", "object"):
+            print(f"  {cql_nullable_types[arg['type']].ljust(20)} {arg['name'].ljust(25)} = NULL;")
+        elif arg['isNotNull']:
+            print(f"  {cql_notnull_types[arg['type']].ljust(20)} {arg['name'].ljust(25)} = 0;")
+        else:
+            print(f"  {cql_nullable_types[arg['type']].ljust(20)} {(arg['name'] + ';').ljust(25)}   cql_set_null({arg['name']});")
+    print("")
+
+
+    print("  // 4. Initialize procedure dependencies")
+    if proc['projection']:   print(f"  {proc['name']}_result_set_ref _data_result_set_ = NULL;")
+    if proc['usesDatabase']: print(f"  sqlite3* db = sqlite3_context_db_handle(context);")
+    if proc['usesDatabase']: print(f"  cql_code rc = SQLITE_OK;")
+    print("") if proc['usesDatabase'] or proc['projection'] else None
+
+
+    print("  // 5. Call the procedure")
+    print(f"  {'rc = ' if proc['usesDatabase'] else ''}{proc['name']}{'_fetch_results' if proc['projection'] else ''}(", end="")
+    for index, computed_arg in enumerate(
```

Comment:
I might hoist the arg building out into a local so that the arg building doesn't make the printing harder to read.

This is all idoimatic python so maybe people are used to reading this.  My python looks like it's C :D